File: 1706-where-will-the-ball-fall/1706-where-will-the-ball-fall.java
class Solution {
    public int[] findBall(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        
        int[] ans = new int[n];
        for(int col = 0; col < n; col++) {
            int r = 0, c = col;
            while(r < m){
                if(grid[r][c] == 1) {
                    if(c == n - 1 || grid[r][c + 1] == -1) {
                        ans[col] = -1;
                        break;
                    } else {
                        r++;
                        c++; 
                    }  
                } else {
                    if(c == 0 || grid[r][c - 1] == 1){
                        ans[col] = -1;
                        break;
                    } else {
                        r++;
                        c--; 
                    }  
                }
            }
            if(ans[col] != -1){
                ans[col] = c;
            }
        }
        return ans;
    }
}

File: 1706-where-will-the-ball-fall/1706-where-will-the-ball-fall.py
class Solution:
    def findBall(self, grid: List[List[int]]) -> List[int]:
        m, n = len(grid), len(grid[0])
        ans = [-1] * n
        for c in range(n):
            j = c
            for r in range(m):
                if grid[r][j] == 1:
                    if j == n - 1 or grid[r][j + 1] == -1:
                        break
                    j += 1
                else:
                    if j == 0 or grid[r][j - 1] == 1:
                        break
                    j -= 1
            else:
                ans[c] = j
        return ans